### PR TITLE
Implement ability to hide warnings in VO table parsing

### LINKED ISSUE #8028: Silence warnings by default when reading in VO Tables
### TL;DR

Users often are given files they don't have control over, and those files aren't always standard-compliant. This is especially true of VO Tables. I'd like to suggest that we make the VO Table reader more forgiving, although the *writer* should continue to emit warnings. Obviously we should discuss this first before doing, but I just want to put the proposal out there.

### Details

Taking the example of VO Tables, the following is an example of reading in one of the files in our test suite (which wasn't there to test warnings):

```
In [5]: parse('gemini.xml')
WARNING: W49: gemini.xml:37:12: W49: Empty cell illegal for integer fields. [astropy.io.votable.converters]
WARNING: W49: gemini.xml:49:12: W49: Empty cell illegal for integer fields. [astropy.io.votable.converters]
WARNING: W49: gemini.xml:61:12: W49: Empty cell illegal for integer fields. [astropy.io.votable.converters]
WARNING: W48: gemini.xml:78:10: W48: Unknown attribute 'value' on OPTION [astropy.io.votable.tree]
WARNING: W48: gemini.xml:79:10: W48: Unknown attribute 'value' on OPTION [astropy.io.votable.tree]
WARNING: W06: gemini.xml:98:6: W06: Invalid UCD 'obs.field': Secondary word 'obs.field' is not valid as a primary word [astropy.io.votable.tree]
WARNING: W06: gemini.xml:99:6: W06: Invalid UCD 'obs.field': Secondary word 'obs.field' is not valid as a primary word [astropy.io.votable.tree]
WARNING: E02: gemini.xml:99:6: E02: Incorrect number of elements in array. Expected multiple of 3, got 1 [astropy.io.votable.converters]
WARNING: W06: gemini.xml:100:6: W06: Invalid UCD 'obs.field': Secondary word 'obs.field' is not valid as a primary word [astropy.io.votable.tree]
WARNING: W06: gemini.xml:101:6: W06: Invalid UCD 'em.wl;stat.interval': Unknown word 'stat.interval' [astropy.io.votable.tree]
WARNING: E02: gemini.xml:101:6: E02: Incorrect number of elements in array. Expected multiple of 2, got 1 [astropy.io.votable.converters]
WARNING: W06: gemini.xml:102:6: W06: Invalid UCD 'time;stat.interval': Unknown word 'stat.interval' [astropy.io.votable.tree]
WARNING: E02: gemini.xml:102:6: E02: Incorrect number of elements in array. Expected multiple of 2, got 1 [astropy.io.votable.converters]
WARNING: W06: gemini.xml:112:6: W06: Invalid UCD 'obs.field': Secondary word 'obs.field' is not valid as a primary word [astropy.io.votable.tree]
WARNING: W06: gemini.xml:113:6: W06: Invalid UCD 'obs.field': Secondary word 'obs.field' is not valid as a primary word [astropy.io.votable.tree]
WARNING: E02: gemini.xml:113:6: E02: Incorrect number of elements in array. Expected multiple of 3, got 1 [astropy.io.votable.converters]
WARNING: W06: gemini.xml:114:6: W06: Invalid UCD 'obs.field': Secondary word 'obs.field' is not valid as a primary word [astropy.io.votable.tree]
WARNING: W06: gemini.xml:115:6: W06: Invalid UCD 'em.wl;stat.interval': Unknown word 'stat.interval' [astropy.io.votable.tree]
WARNING: E02: gemini.xml:115:6: E02: Incorrect number of elements in array. Expected multiple of 2, got 1 [astropy.io.votable.converters]
WARNING: W06: gemini.xml:116:6: W06: Invalid UCD 'time;stat.interval': Unknown word 'stat.interval' (suppressing further warnings of this type...) [astropy.io.votable.tree]
WARNING: E02: gemini.xml:116:6: E02: Incorrect number of elements in array. Expected multiple of 2, got 1 [astropy.io.votable.converters]
WARNING: E02: gemini.xml:127:6: E02: Incorrect number of elements in array. Expected multiple of 3, got 1 [astropy.io.votable.converters]
WARNING: E02: gemini.xml:137:6: E02: Incorrect number of elements in array. Expected multiple of 2, got 1 [astropy.io.votable.converters]
WARNING: E02: gemini.xml:151:6: E02: Incorrect number of elements in array. Expected multiple of 3, got 1 [astropy.io.votable.converters]
WARNING: E02: gemini.xml:161:6: E02: Incorrect number of elements in array. Expected multiple of 2, got 1 (suppressing further warnings of this type...) [astropy.io.votable.converters]
Out[5]: <VOTABLE>... 1 tables ...</VOTABLE>
```

This is a pretty typical number of warnings in my experience with VO Tables. I've never done anything about any of the warnings though...

Note that there is actually a way to be even more pedantic:

```
In [6]: parse('gemini.xml', pedantic=True)
---------------------------------------------------------------------------
W49                                       Traceback (most recent call last)
<ipython-input-6-70047e7af5ca> in <module>()
----> 1 parse('gemini.xml', pedantic=True)

~/Dropbox/Code/Astropy/astropy/astropy/io/votable/table.py in parse(source, columns, invalid, pedantic, chunk_size, table_number, table_id, filename, unit_format, datatype_mapping, _debug_python_based_parser)
    135             _debug_python_based_parser=_debug_python_based_parser) as iterator:
    136         return tree.VOTableFile(
--> 137             config=config, pos=(1, 1)).parse(iterator, config)
    138 
    139 
...
~/Dropbox/Code/Astropy/astropy/astropy/io/votable/exceptions.py in vo_raise(exception_class, args, config, pos)
     96     if config is None:
     97         config = {}
---> 98     raise exception_class(args, config, pos)
     99 
    100 

W49: gemini.xml:37:12: W49: Empty cell illegal for integer fields.
```

But actually no way to be less pedantic and ignore the warnings (short of using ``warnings.catch_warnigns``. I'd like to suggest that we add a ``verify`` key to the VO Table ``parse`` which can take different options as for FITS, including ``ignore``, ``warn``, ``exception`` (and possibly deprecate ``pendantic``).

Furthermore, I think we might want to consider defaulting to ``'ignore'``.

We could also do something similar with FITS files - ignore warnings when reading but show them when writing?

### PR DESCRIPTION
**Pull request by astrofrog**
This PR makes it possible to hide warnings when parsing VO tables, by replacing the boolean ``pedantic`` flag with a ``verify`` flag that can be ``ignore``, ``warn``, or ``exception`` (or should it be ``error``?)

In addition, as discussed in https://github.com/astropy/astropy/issues/8028, this makes ``ignore`` the default. I think in general we should move to a model where we don't give 25 warnings when reading in a table where these are fixed anyway and beyond the user's control (basically the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle) of being liberal in what you accept)

I've tagged various people I think might be interested. If we go ahead with this I think in another PR we should also change the default for FITS to not show fixable warnings on input.

EDIT: Fix #8028

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -112,6 +112,11 @@ astropy.io.registry
 astropy.io.votable
 ^^^^^^^^^^^^^^^^^^
 
+- Changed ``pedantic`` argument to ``verify`` and change it to have three
```
Comment (bsipocz) [inline CHANGES.rst]: I think this should go into the AP changes section

File: astropy/visualization/wcsaxes/core.py [hunk 2]
```diff
@@ -398,6 +398,8 @@ def reset_wcs(self, wcs=None, slices=None, transform=None, coord_meta=None):
 
     def draw_wcsaxes(self, renderer):
 
+        print("HERE")
```
Comment (bsipocz) [inline astropy/visualization/wcsaxes/core.py]: this changes are not related to this PR

**Comment by pllim** [comment]
> or should it be `error`?

Naw, it is also `exception` on `io.fits` (http://docs.astropy.org/en/stable/io/fits/usage/verification.html), so in the spirit of consistency, we can use that.

File: CHANGES.rst (hunk 1 above)
Comment (astrofrog) [reply CHANGES.rst]: I thought it was already?

File: astropy/visualization/wcsaxes/core.py (hunk 2 above)
Comment (astrofrog) [reply astropy/visualization/wcsaxes/core.py]: Fixed

**Comment by astrofrog** [comment]
@bsipocz:

> Also I would think that adding a 'fix' option for the same kwarg would be very nice, that would also nicely mirror the behaviour/options of the fits .verify()

This would require a lot more work since at the moment there is no control over whether things are fixed or not. They are always fixed, and it could be that the VO table can't be interpreted if not fixed (a little different to FITS headers which can be sub-optimal but still work). I'd be tempted to keep just these three options for now, and then consider adding more in future if people really want?

File: astropy/io/votable/table.py [hunk 3]
```diff
@@ -17,15 +17,19 @@
...
 
 
 __all__ = ['parse', 'parse_single_table', 'from_table', 'writeto', 'validate',
            'reset_vo_warnings']
 
+VERIFY_OPTIONS = ['ignore', 'warn', 'exception']
 
+
+@deprecated_renamed_argument('pedantic', 'verify', pending=True, since='4.0')
 def parse(source, columns=None, invalid='exception', pedantic=None,
```
Comment (pllim) [inline astropy/io/votable/table.py]:
```suggestion
def parse(source, columns=None, invalid='exception', verify=None,
```

File: astropy/io/votable/table.py [hunk 4]
```diff
@@ -49,12 +53,17 @@ def parse(source, columns=None, invalid='exception', pedantic=None,
             - 'mask': mask out invalid values
 
     pedantic : bool, optional
-        When `True`, raise an error when the file violates the spec,
-        otherwise issue a warning.  Warnings may be controlled using
-        the standard Python mechanisms.  See the `warnings`
-        module in the Python standard library for more information.
-        When not provided, uses the configuration setting
-        ``astropy.io.votable.pedantic``, which defaults to False.
+        This is now an alias for ``verify='warn'`` (if ``pedantic`` is `False`)
```
Comment (pllim) [inline astropy/io/votable/table.py]: Not sure if `pedantic` should be documented in docstring at all, according to precedence in #5171 .

File: astropy/io/votable/table.py [hunk 5]
```diff
@@ -110,16 +119,24 @@ def parse(source, columns=None, invalid='exception', pedantic=None,
         raise ValueError("accepted values of ``invalid`` are: "
                          "``'exception'`` or ``'mask'``.")
 
-    if pedantic is None:
-        pedantic = conf.pedantic
+    if verify is None:
```
Comment (pllim) [inline astropy/io/votable/table.py]: Don't need this?

File: astropy/io/votable/table.py [hunk 6]
```diff
@@ -49,12 +53,17 @@ def parse(source, columns=None, invalid='exception', pedantic=None,
...
+        or ``verify='exception'`` (if ``pedantic`` is `True`). The ``pedantic``
+        option will be deprecated in future.
+
+    verify : {'ignore', 'warn', 'exception'}, optional
+        When ``'exception'``, raise an error when the file violates the spec,
+        otherwise either issue a warning (``'warn'``) or silently continue
+        (``'ignore'``). Warnings may be controlled using the standard Python
+        mechanisms.  See the `warnings` module in the Python standard library
+        for more information. When not provided, uses the configuration setting
+        ``astropy.io.votable.verify``, which defaults to 'ignore'.
```
Comment (pllim) [inline astropy/io/votable/table.py]: Need to add a note of `.. versionchanged::`

File: astropy/io/votable/tree.py [hunk 7]
```diff
@@ -1170,7 +1172,7 @@ def __init__(self, votable, ID=None, name=None, datatype=None,
         # actually contains character data.  We have to hack the field
         # to store character data, or we can't read it in.  A warning
         # will be raised when this happens.
-        if (not config.get('pedantic') and name == 'cprojection' and
+        if (config.get('verify') == 'warn' and name == 'cprojection' and
```
Comment (pllim) [inline astropy/io/votable/tree.py]: With this change, this `if` block is also skipped when `verify` is `'ignore'`, in which `datatype` and `arraysize` are also not set. Will this introduce a bug?

**Comment by bsipocz** [comment]
@astrofrog 
> Also I would think that adding a 'fix' option for the same kwarg would be very nice, that would also nicely mirror the behaviour/options of the fits .verify()

out of band discussion showed me why this is not a good idea, so please ignore it, it shouldn't actually fix anything.

File: astropy/io/votable/__init__.py [hunk 8]
```diff
@@ -24,10 +24,13 @@ class Conf(_config.ConfigNamespace):
     Configuration parameters for `astropy.io.votable`.
     """
 
-    pedantic = _config.ConfigItem(
-        False,
-        'When True, treat fixable violations of the VOTable spec as exceptions.',
-        aliases=['astropy.io.votable.table.pedantic'])
+    verify = _config.ConfigItem(
+        'ignore',
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/votable/__init__.py]: For some reason it still doesn't work right if I specify ``option`` with mixed types, and it always gets converted to a string. But that's ok, we can still be backward compatible by just checking for False and True as strings when using ``conf.verify``. I've added handling of this as well as an extensive test of setting the configuration items (both verify and pedantic).

File: astropy/io/votable/exceptions.py [hunk 9]
```diff
@@ -122,8 +122,9 @@ def vo_warn(warning_class, args=(), config=None, pos=None, stacklevel=1):
     """
     if config is None:
         config = {}
-    warning = warning_class(args, config, pos)
-    _suppressed_warning(warning, config, stacklevel=stacklevel+1)
+    if config.get('verify') != 'ignore':
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/votable/exceptions.py]: Fixed

File: astropy/io/votable/exceptions.py [hunk 10]
```diff
@@ -430,8 +431,8 @@ class W09(VOTableSpecWarning):
     The VOTable specification uses the attribute name ``ID`` (with
     uppercase letters) to specify unique identifiers.  Some
     VOTable-producing tools use the more standard lowercase ``id``
-    instead.  ``vo.table`` accepts ``id`` and emits this warning when
-    not in ``pedantic`` mode.
+    instead.  ``vo.table`` accepts ``id`` and emits this warning if
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/votable/exceptions.py]: Fixed

File: astropy/io/votable/table.py (hunk 4 above)
Comment (astrofrog) [reply astropy/io/votable/table.py]: Fixed

File: astropy/io/votable/table.py (hunk 5 above)
Comment (astrofrog) [reply astropy/io/votable/table.py]: Fixed

File: astropy/io/votable/table.py (hunk 6 above)
Comment (astrofrog) [reply astropy/io/votable/table.py]: Fixed

File: astropy/io/votable/tree.py (hunk 7 above)
Comment (astrofrog) [reply astropy/io/votable/tree.py]: Fixed

File: astropy/io/votable/converters.py [hunk 11]
```diff
@@ -160,7 +160,7 @@ class Converter:
     """
 
     def __init__(self, field, config=None, pos=None):
-        pass
+        self._config = config
```
Comment (astrofrog) [inline astropy/io/votable/converters.py]: This should fix what I think is a real bug that didn't manifest itself before due to the different default of emitting warnings. This behavior makes sense anyway - if a config is passed to the initializer, it should be the default for the method calls after unless overridden.

**Comment by astrofrog** [comment]
@pllim - I've implemented all your comments. Feel free to take this for a spin!

File: astropy/io/votable/converters.py (hunk 11 above)
Comment (astrofrog) [reply astropy/io/votable/converters.py]: Actually I reverted this since it's not actually needed anymore

File: astropy/io/votable/table.py [hunk 12]
```diff
@@ -110,16 +118,27 @@ def parse(source, columns=None, invalid='exception', pedantic=None,
...
+    if verify is None:
+        # Note that we need to allow verify to be booleans as strings since
+        # the configuration framework doesn't make it easy/possible to have
+        # mixed types.
+        if conf.verify.lower() in ['false', 'true']:
+            verify = conf.verify.lower() == 'true'
+        else:
+            verify = conf.verify
+
+    if isinstance(verify, bool):
```
Comment (pllim) [inline astropy/io/votable/table.py]: Extra check here no longer necessary after the simplification above.
```suggestion
```

File: astropy/io/votable/table.py [hunk 13]
```diff
@@ -110,16 +118,27 @@ def parse(source, columns=None, invalid='exception', pedantic=None,
...
+        verify = 'exception' if verify else 'warn'
```
Comment (pllim) [inline astropy/io/votable/table.py]:
```suggestion
```

File: astropy/io/votable/table.py [hunk 14]
```diff
@@ -110,16 +118,27 @@ def parse(source, columns=None, invalid='exception', pedantic=None,
...
+    elif verify not in VERIFY_OPTIONS:
```
Comment (pllim) [inline astropy/io/votable/table.py]:
```suggestion
    if verify not in VERIFY_OPTIONS:
```

File: astropy/io/votable/tests/table_test.py [hunk 15]
```diff
@@ -175,6 +179,80 @@ def test_write_with_format():
...
+    assert len(w) == 25
+
+    with pytest.raises(VOWarning):
+        parse(get_pkg_data_filename('data/gemini.xml'), verify='exception')
+
+    # Make sure the pedantic option still works for now (pending deprecation)
+
+    with catch_warnings(VOWarning) as w:
+        parse(get_pkg_data_filename('data/gemini.xml'), pedantic=False)
+    assert len(w) == 25
```
Comment (pllim) [inline astropy/io/votable/tests/table_test.py]: For the `pedantic` backward-compat test, we should also check that `AstropyDeprecationWarning` is not *yet* raised. This way, when we remove `pending=True`, the tests will fail and we will have to update the test to catch the deprecation warning, and the warning wouldn't "leak through" in a way that `pytest` would complain ungracefully about it.

File: astropy/io/votable/tests/table_test.py [hunk 16]
```diff
@@ -175,6 +179,80 @@ def test_write_with_format():
...
+
+        with open(tmpdir.join('astropy').join('astropy.cfg').strpath, 'w') as f:
+            f.write('[io.votable]\npedantic = True')
+
+        reload_config('astropy.io.votable')
+
+        with pytest.raises(VOWarning):
+            parse(get_pkg_data_filename('data/gemini.xml'))
+
+    reload_config('astropy.io.votable')
```
Comment (pllim) [inline astropy/io/votable/tests/table_test.py]: What does this one do, exactly?

File: astropy/io/votable/tests/table_test.py [hunk 17]
```diff
@@ -175,6 +179,80 @@ def test_write_with_format():
 def test_empty_table():
     votable = parse(
         get_pkg_data_filename('data/empty_table.xml'),
-        pedantic=False)
+        verify='warn')
     table = votable.get_first_table()
     astropy_table = table.to_table()  # noqa
+
+
+def test_verify_options(tmpdir):
```
Comment (pllim) [inline astropy/io/votable/tests/table_test.py]: I feel like this can be turned into a test class with the different tests running separately as methods. This way, when one of them fail, it will be clearer *which one* is failing.

File: astropy/io/votable/tests/table_test.py (hunk 16 above)
Comment (astrofrog) [reply astropy/io/votable/tests/table_test.py]: Once we exit the context manager, the config dir changes back to the original directory so we need to reload again to get the original config back, I think. I'll double check and can add a comment.

File: astropy/io/votable/tests/table_test.py (hunk 17 above)
Comment (astrofrog) [reply astropy/io/votable/tests/table_test.py]: Will do!

File: astropy/io/votable/tests/table_test.py (hunk 16 above)
Comment (astrofrog) [reply astropy/io/votable/tests/table_test.py]: But this is no longer needed once the tests are refactored into a class...

File: astropy/io/votable/tests/table_test.py (hunk 15 above)
Comment (astrofrog) [reply astropy/io/votable/tests/table_test.py]: Done

File: astropy/io/votable/tests/table_test.py (hunk 17 above)
Comment (astrofrog) [reply astropy/io/votable/tests/table_test.py]: Done

**Comment by astrofrog** [comment]
@pllim - I had to change some of the code in ``parse`` back to being able to handle booleans, because we need to catch the case where the user calls the function with ``pedantic=True`` for instance. Otherwise I think I've accepted/implemented all your suggestions!

File: astropy/io/votable/tests/table_test.py [hunk 18]
```diff
@@ -175,6 +180,93 @@ def test_write_with_format():
...
+        assert len(w) == 25
+
+    def test_verify_exception(self):
+        with pytest.raises(VOWarning):
+            parse(get_pkg_data_filename('data/gemini.xml'), verify='exception')
+
+    # Make sure the pedantic option still works for now (pending deprecation)
+
+    def test_pedantic_false(self):
+        with catch_warnings(VOWarning) as w:
```
Comment (pllim) [inline astropy/io/votable/tests/table_test.py]: To check for it below, you first need to catch it. (Gotta catch them all... Pokemon!)
```suggestion
        with catch_warnings(VOWarning, AstropyDeprecationWarning) as w:
```

File: astropy/io/votable/tests/table_test.py [hunk 19]
```diff
@@ -175,6 +180,93 @@ def test_write_with_format():
...
+    def test_conf_pedantic_false(self, tmpdir):
+
+        with set_temp_config(tmpdir.strpath):
+
+            with open(tmpdir.join('astropy').join('astropy.cfg').strpath, 'w') as f:
+                f.write('[io.votable]\npedantic = False')
+
+            reload_config('astropy.io.votable')
+
+            with catch_warnings(VOWarning) as w:
```
Comment (pllim) [inline astropy/io/votable/tests/table_test.py]:
```suggestion
            with catch_warnings(VOWarning, AstropyDeprecationWarning) as w:
```

File: astropy/io/votable/tests/vo_test.py [hunk 20]
```diff
@@ -50,7 +50,7 @@ def assert_validate_schema(filename, version):
 def test_parse_single_table():
     table = parse_single_table(
         get_pkg_data_filename('data/regression.xml'),
-        pedantic=False)
+        verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 21]
```diff
@@ -59,7 +59,7 @@ def test_parse_single_table2():
     table2 = parse_single_table(
         get_pkg_data_filename('data/regression.xml'),
         table_number=1,
-        pedantic=False)
+        verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 22]
```diff
@@ -69,14 +69,14 @@ def test_parse_single_table2():
 def test_parse_single_table3():
     parse_single_table(
         get_pkg_data_filename('data/regression.xml'),
-        table_number=3, pedantic=False)
+        table_number=3, verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        table_number=3, verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 23]
```diff
@@ -69,14 +69,14 @@ def test_parse_single_table2():
...
 
 
 def _test_regression(tmpdir, _python_based=False, binary_mode=1):
     # Read the VOTABLE
     votable = parse(
         get_pkg_data_filename('data/regression.xml'),
-        pedantic=False,
+        verify='warn',
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore',
```

File: astropy/io/votable/tests/vo_test.py [hunk 24]
```diff
@@ -139,7 +139,7 @@ def _test_regression(tmpdir, _python_based=False, binary_mode=1):
                            votable.version)
     # Also try passing a file handle
     with open(str(tmpdir.join("regression.binary.xml")), "rb") as fd:
-        votable2 = parse(fd, pedantic=False,
+        votable2 = parse(fd, verify='warn',
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        votable2 = parse(fd, verify='ignore',
```

File: astropy/io/votable/tests/vo_test.py [hunk 25]
```diff
@@ -198,7 +198,7 @@ class TestFixups:
     def setup_class(self):
         self.table = parse(
             get_pkg_data_filename('data/regression.xml'),
-            pedantic=False).get_first_table()
+            verify='warn').get_first_table()
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore').get_first_table()
```

File: astropy/io/votable/tests/vo_test.py [hunk 26]
```diff
@@ -211,7 +211,7 @@ class TestReferences:
     def setup_class(self):
         self.votable = parse(
             get_pkg_data_filename('data/regression.xml'),
-            pedantic=False)
+            verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 27]
```diff
@@ -252,7 +252,7 @@ def test_select_columns_by_index():
     columns = [0, 5, 13]
     table = parse(
         get_pkg_data_filename('data/regression.xml'),
-        pedantic=False, columns=columns).get_first_table()
+        verify='warn', columns=columns).get_first_table()
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore', columns=columns).get_first_table()
```

File: astropy/io/votable/tests/vo_test.py [hunk 28]
```diff
@@ -266,7 +266,7 @@ def test_select_columns_by_name():
     columns = ['string_test', 'unsignedByte', 'bitarray']
     table = parse(
         get_pkg_data_filename('data/regression.xml'),
-        pedantic=False, columns=columns).get_first_table()
+        verify='warn', columns=columns).get_first_table()
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore', columns=columns).get_first_table()
```

File: astropy/io/votable/tests/vo_test.py [hunk 29]
```diff
@@ -279,7 +279,7 @@ class TestParse:
     def setup_class(self):
         self.votable = parse(
             get_pkg_data_filename('data/regression.xml'),
-            pedantic=False)
+            verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 30]
```diff
@@ -611,12 +611,12 @@ class TestThroughTableData(TestParse):
     def setup_class(self):
         votable = parse(
             get_pkg_data_filename('data/regression.xml'),
-            pedantic=False)
+            verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 31]
```diff
@@ -611,12 +611,12 @@ class TestThroughTableData(TestParse):
...
 
         self.xmlout = bio = io.BytesIO()
         votable.to_xml(bio)
         bio.seek(0)
-        self.votable = parse(bio, pedantic=False)
+        self.votable = parse(bio, verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        self.votable = parse(bio, verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 32]
```diff
@@ -644,13 +644,13 @@ class TestThroughBinary(TestParse):
     def setup_class(self):
         votable = parse(
             get_pkg_data_filename('data/regression.xml'),
-            pedantic=False)
+            verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 33]
```diff
@@ -644,13 +644,13 @@ class TestThroughBinary(TestParse):
...
         votable.get_first_table().format = 'binary'
 
         self.xmlout = bio = io.BytesIO()
         votable.to_xml(bio)
         bio.seek(0)
-        self.votable = parse(bio, pedantic=False)
+        self.votable = parse(bio, verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        self.votable = parse(bio, verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 34]
```diff
@@ -673,15 +673,15 @@ class TestThroughBinary2(TestParse):
     def setup_class(self):
         votable = parse(
             get_pkg_data_filename('data/regression.xml'),
-            pedantic=False)
+            verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 35]
```diff
@@ -673,15 +673,15 @@ class TestThroughBinary2(TestParse):
...
         votable.version = '1.3'
         votable.get_first_table()._config['version_1_3_or_later'] = True
         votable.get_first_table().format = 'binary2'
 
         self.xmlout = bio = io.BytesIO()
         votable.to_xml(bio)
         bio.seek(0)
-        self.votable = parse(bio, pedantic=False)
+        self.votable = parse(bio, verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        self.votable = parse(bio, verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 36]
```diff
@@ -729,14 +729,14 @@ def test_open_files():
     for filename in get_pkg_data_filenames('data', pattern='*.xml'):
         if filename.endswith('custom_datatype.xml'):
             continue
-        parse(filename, pedantic=False)
+        parse(filename, verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        parse(filename, verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 37]
```diff
@@ -729,14 +729,14 @@ def test_open_files():
...
 
 
 @raises(VOTableSpecError)
 def test_too_many_columns():
     parse(
         get_pkg_data_filename('data/too_many_columns.xml.gz'),
-        pedantic=False)
+        verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 38]
```diff
@@ -839,7 +839,7 @@ def test_validate_path_object():
 def test_gzip_filehandles(tmpdir):
     votable = parse(
         get_pkg_data_filename('data/regression.xml'),
-        pedantic=False)
+        verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 39]
```diff
@@ -850,7 +850,7 @@ def test_gzip_filehandles(tmpdir):
     with open(str(tmpdir.join("regression.compressed.xml")), 'rb') as fd:
         votable = parse(
             fd,
-            pedantic=False)
+            verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
            verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 40]
```diff
@@ -910,14 +910,14 @@ def test_nonstandard_units():
 
     votable = parse(
         get_pkg_data_filename('data/nonstandard_units.xml'),
-        pedantic=False)
+        verify='warn')
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore')
```

File: astropy/io/votable/tests/vo_test.py [hunk 41]
```diff
@@ -910,14 +910,14 @@ def test_nonstandard_units():
...
 
     assert isinstance(
         votable.get_first_table().fields[0].unit, u.UnrecognizedUnit)
 
     votable = parse(
         get_pkg_data_filename('data/nonstandard_units.xml'),
-        pedantic=False,
+        verify='warn',
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore',
```

File: astropy/io/votable/tests/vo_test.py [hunk 42]
```diff
@@ -1012,7 +1012,7 @@ def test_instantiate_vowarning():
 def test_custom_datatype():
     votable = parse(
         get_pkg_data_filename('data/custom_datatype.xml'),
-        pedantic=False,
+        verify='warn',
```
Comment (pllim) [inline astropy/io/votable/tests/vo_test.py]:
```suggestion
        verify='ignore',
```

File: astropy/io/votable/tree.py [hunk 43]
```diff
@@ -268,11 +268,13 @@ def check_ucd(ucd, config=None, pos=None):
                 has_colon=config.get('version_1_2_or_later', False))
         except ValueError as e:
             # This weird construction is for Python 3 compatibility
-            if config.get('pedantic'):
+            if config.get('verify') == 'exception':
```
Comment (pllim) [inline astropy/io/votable/tree.py]:
```suggestion
            if config.get('verify', 'ignore') == 'exception':
```

File: astropy/io/votable/tree.py [hunk 44]
```diff
@@ -268,11 +268,13 @@ def check_ucd(ucd, config=None, pos=None):
...
                 vo_raise(W06, (ucd, str(e)), config, pos)
-            else:
+            elif config.get('verify') == 'warn':
```
Comment (pllim) [inline astropy/io/votable/tree.py]:
```suggestion
            elif config.get('verify', 'ignore') == 'warn':
```

File: astropy/io/votable/tree.py [hunk 45]
```diff
@@ -1170,7 +1172,7 @@ def __init__(self, votable, ID=None, name=None, datatype=None,
         # actually contains character data.  We have to hack the field
         # to store character data, or we can't read it in.  A warning
         # will be raised when this happens.
-        if (not config.get('pedantic') and name == 'cprojection' and
+        if (config.get('verify') != 'exception' and name == 'cprojection' and
```
Comment (pllim) [inline astropy/io/votable/tree.py]:
```suggestion
        if (config.get('verify', 'ignore') != 'exception' and name == 'cprojection' and
```

**Comment by pllim** [comment]
p.s. I am excited to see how many warnings disappear in #7928 once this is merged and I rebase that one on top of this one!

**Comment by pllim** [comment]
Hmm, I don't understand this one:

```
            with catch_warnings(VOWarning, AstropyDeprecationWarning) as w:
                parse(get_pkg_data_filename('data/gemini.xml'))
>           assert len(w) == 25
E           assert 26 == 25
E            +  where 26 = len([<warnings.WarningMessage object at 0x7fcefa18ec18>, <warnings.WarningMessage object at 0x7fcefa1e8b00>, <warnings.War...fa1e8cc0>, <warnings.WarningMessage object at 0x7fcefa1e8c18>, <warnings.WarningMessage object at 0x7fcefa1e8128>, ...])
```

`pending=True` would have skipped the warning altogether, so why is there an extra warning caught?

https://github.com/astropy/astropy/blob/23e906912453d14f2ecb8f24c1e9b73fdc11c768/astropy/utils/decorators.py#L482-L484

**Comment by astrofrog** [comment]
The issue is that ``ConfigItem`` doesn't have the option to emit pending deprecation warnings, so I've just patched up ``parse`` to ensure that we don't yet emit the deprecation warning when accessing the configuration item.

I've also decided to remove any explicit ``verify='ignore'`` in tests (that don't relate to the verify option) since this is the default, and it makes the code more concise in many cases.

**Comment by astrofrog** [comment]
(rebased to get rid of the docs build error)

**Comment by pllim** [comment]
:shipit:
